File: package/umt_python/src/number/format_number.py
"""Locale-aware number formatting ported from package/main Number/formatNumber.ts."""

from dataclasses import dataclass
from typing import Literal

FormatStyle = Literal["decimal", "currency", "percent"]


@dataclass(frozen=True)
class _LocaleSpec:
    """Separator characters used when formatting numbers for a locale."""

    thousands: str
    decimal: str


# Separator tables for the handful of locales we explicitly support. Any
# unknown locale falls back to the en-US layout below, which mirrors the
# default behavior of `Intl.NumberFormat` when the ICU data is missing.
_LOCALE_SEPARATORS: dict[str, _LocaleSpec] = {
    "en-US": _LocaleSpec(thousands=",", decimal="."),
    "en-GB": _LocaleSpec(thousands=",", decimal="."),
    "ja-JP": _LocaleSpec(thousands=",", decimal="."),
    "de-DE": _LocaleSpec(thousands=".", decimal=","),
    "fr-FR": _LocaleSpec(thousands="\u202f", decimal=","),
    "it-IT": _LocaleSpec(thousands=".", decimal=","),
    "es-ES": _LocaleSpec(thousands=".", decimal=","),
}

_DEFAULT_LOCALE_SPEC = _LOCALE_SEPARATORS["en-US"]

# Minimal symbol table so the "currency" style produces a useful prefix
# without pulling in a full ICU / Babel dependency (the Python package is
# zero-dependency by policy, matching package/main). Unknown currencies
# fall back to the ISO 4217 code plus a non-breaking space.
_CURRENCY_SYMBOLS: dict[str, str] = {
    "USD": "$",
    "EUR": "\u20ac",
    "GBP": "\u00a3",
    "JPY": "\u00a5",
    "CNY": "\u00a5",
    "KRW": "\u20a9",
}


def _resolve_locale_spec(locale: str | None) -> _LocaleSpec:
    if locale is None:
        return _DEFAULT_LOCALE_SPEC
    # Exact match first, then language-only fallback (e.g. "fr" -> "fr-FR").
    if locale in _LOCALE_SEPARATORS:
        return _LOCALE_SEPARATORS[locale]
    language = locale.split("-", 1)[0]
    for key, spec in _LOCALE_SEPARATORS.items():
        if key.split("-", 1)[0] == language:
            return spec
    return _DEFAULT_LOCALE_SPEC


def _default_fraction_digits(
    style: FormatStyle,
    currency: str | None,
) -> tuple[int, int]:
    """Mirror Intl.NumberFormat defaults for minimum/maximum fraction digits."""

    if style == "currency":
        # JPY and KRW historically have no fractional unit; everything else
        # defaults to 2 digits per ECMA-402.
        if currency in {"JPY", "KRW"}:
            return (0, 0)
        return (2, 2)
    if style == "percent":
        return (0, 0)
    # decimal (default)
    return (0, 3)


def _round_half_away_from_zero(value: float, digits: int) -> float:
    """Round matching Intl.NumberFormat (half away from zero)."""

    if digits < 0:
        msg = "digits must be non-negative"
        raise ValueError(msg)
    factor = 10**digits
    scaled = value * factor
    # Python's round() uses banker's rounding, which diverges from the
    # half-away-from-zero behavior used by Intl.NumberFormat. Emulate the
    # JS semantics so the port matches package/main.
    rounded = int(scaled + 0.5) if scaled >= 0 else -int(-scaled + 0.5)
    return rounded / factor


def _format_absolute(
    value: float,
    min_fraction: int,
    max_fraction: int,
    spec: _LocaleSpec,
) -> str:
    rounded = _round_half_away_from_zero(value, max_fraction)
    # Split into integer and fractional parts via string manipulation so we
    # can apply locale-specific separators independently.
    integer_part = int(rounded)
    remainder = abs(rounded - integer_part)

    integer_str = f"{integer_part:,}".replace(",", spec.thousands)

    if max_fraction == 0:
        return integer_str

    # Produce exactly max_fraction digits then trim trailing zeros down to
    # min_fraction, matching Intl.NumberFormat's default behavior.
    fraction_digits = f"{remainder:.{max_fraction}f}"[2:]
    trimmed = fraction_digits.rstrip("0")
    if len(trimmed) < min_fraction:
        trimmed = fraction_digits[:min_fraction]
    if not trimmed:
        return integer_str
    return f"{integer_str}{spec.decimal}{trimmed}"


def format_number(
    value: float | int,
    *,
    locale: str | None = None,
    minimum_fraction_digits: int | None = None,
    maximum_fraction_digits: int | None = None,
    style: FormatStyle = "decimal",
    currency: str | None = None,
) -> str:
    """Format a number similarly to ``Intl.NumberFormat`` in the TypeScript source.

    Supports decimal, percent, and currency styles, with configurable
    minimum/maximum fraction digits and locale-aware separators for a
    curated list of locales. Unknown locales fall back to ``en-US``.

    Args:
        value: Numeric value to format.
        locale: BCP 47 locale tag (e.g. ``"en-US"``, ``"de-DE"``). Defaults
            to ``"en-US"`` semantics when ``None``.
        minimum_fraction_digits: Minimum number of fractional digits.
        maximum_fraction_digits: Maximum number of fractional digits.
        style: One of ``"decimal"``, ``"currency"``, or ``"percent"``.
        currency: ISO 4217 code used when ``style`` is ``"currency"``.

    Returns:
        The formatted number as a string.

    Example:
        >>> format_number(1234567.89)
        '1,234,567.89'
        >>> format_number(1234567.89, locale="de-DE")
        '1.234.567,89'
        >>> format_number(0.75, style="percent")
        '75%'
        >>> format_number(1234.5, style="currency", currency="USD", locale="en-US")
        '$1,234.50'
    """

    default_min, default_max = _default_fraction_digits(style, currency)
    min_fraction = (
        default_min if minimum_fraction_digits is None else minimum_fraction_digits
    )
    max_fraction = (
        default_max if maximum_fraction_digits is None else maximum_fraction_digits
    )
    # Intl.NumberFormat normalizes this by treating max as the larger of
    # the two \u2014 replicate the same forgiving behavior instead of raising.
    max_fraction = max(max_fraction, min_fraction)

    spec = _resolve_locale_spec(locale)

    scaled_value = value * 100 if style == "percent" else value
    sign = "-" if scaled_value < 0 else ""
    magnitude = _format_absolute(abs(scaled_value), min_fraction, max_fraction, spec)

    if style == "percent":
        return f"{sign}{magnitude}%"

    if style == "currency":
        symbol = _CURRENCY_SYMBOLS.get(
            currency or "", f"{currency}\u00a0" if currency else ""
        )
        return f"{sign}{symbol}{magnitude}"

    return f"{sign}{magnitude}"

File: package/umt_rust/src/number/format_number.rs
// Locale-aware number formatting ported from package/main Number/formatNumber.ts.
//
// The TypeScript reference delegates to `Intl.NumberFormat`, which Rust lacks.
// We reproduce the same observable output for the locales and currencies that
// the Python port (`package/umt_python/src/number/format_number.py`) already
// covers, keeping the three packages in sync.

/// Formatting style accepted by [`umt_format_number`].
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub enum FormatStyle {
    #[default]
    Decimal,
    Currency,
    Percent,
}

/// Options mirroring the TypeScript `FormatNumberOptions` interface.
///
/// Using `Option` for each field allows callers to rely on
/// `Intl.NumberFormat`-style defaults while still overriding individual knobs.
#[derive(Debug, Clone, Default)]
pub struct FormatNumberOptions<'a> {
    /// BCP 47 locale tag (e.g. "en-US", "de-DE"). Defaults to "en-US" semantics.
    pub locale: Option<&'a str>,
    /// Minimum number of fractional digits.
    pub minimum_fraction_digits: Option<u32>,
    /// Maximum number of fractional digits.
    pub maximum_fraction_digits: Option<u32>,
    /// Formatting style (decimal / currency / percent).
    pub style: Option<FormatStyle>,
    /// ISO 4217 code used when `style` is `Currency`.
    pub currency: Option<&'a str>,
}

struct LocaleSpec {
    thousands: &'static str,
    decimal: &'static str,
}

const EN_SPEC: LocaleSpec = LocaleSpec {
    thousands: ",",
    decimal: ".",
};
const DE_SPEC: LocaleSpec = LocaleSpec {
    thousands: ".",
    decimal: ",",
};
const FR_SPEC: LocaleSpec = LocaleSpec {
    thousands: "\u{202f}",
    decimal: ",",
};

fn resolve_locale_spec(locale: Option<&str>) -> LocaleSpec {
    match locale {
        None => EN_SPEC,
        Some(tag) => {
            // Explicit match first, then language-only fallback, then en-US.
            match tag {
                "en-US" | "en-GB" | "ja-JP" => EN_SPEC,
                "de-DE" | "it-IT" | "es-ES" => DE_SPEC,
                "fr-FR" => FR_SPEC,
                _ => match tag.split('-').next().unwrap_or("") {
                    "en" | "ja" => EN_SPEC,
                    "de" | "it" | "es" => DE_SPEC,
                    "fr" => FR_SPEC,
                    _ => EN_SPEC,
                },
            }
        }
    }
}

fn currency_symbol(code: &str) -> Option<&'static str> {
    // Matches the symbol table used by the Python port.
    match code {
        "USD" => Some("$"),
        "EUR" => Some("\u{20ac}"),
        "GBP" => Some("\u{00a3}"),
        "JPY" | "CNY" => Some("\u{00a5}"),
        "KRW" => Some("\u{20a9}"),
        _ => None,
    }
}

// Currencies whose Intl defaults are 0 fraction digits rather than 2.
fn is_zero_fraction_currency(code: &str) -> bool {
    matches!(code, "JPY" | "KRW")
}

fn default_fraction_digits(style: FormatStyle, currency: Option<&str>) -> (u32, u32) {
    match style {
        FormatStyle::Currency => {
            if currency.is_some_and(is_zero_fraction_currency) {
                (0, 0)
            } else {
                (2, 2)
            }
        }
        FormatStyle::Percent => (0, 0),
        FormatStyle::Decimal => (0, 3),
    }
}

// Intl.NumberFormat rounds half away from zero; f64::round matches that
// semantic, unlike the usual banker's rounding that format! applies.
fn round_half_away_from_zero(value: f64, digits: u32) -> f64 {
    let factor = 10_f64.powi(digits as i32);
    (value * factor).round() / factor
}

fn insert_thousands(integer: u128, sep: &str) -> String {
    let digits = integer.to_string();
    let bytes = digits.as_bytes();
    let n = bytes.len();
    let mut out = String::with_capacity(n + n / 3 * sep.len());
    for (i, byte) in bytes.iter().enumerate() {
        if i > 0 && (n - i).is_multiple_of(3) {
            out.push_str(sep);
        }
        out.push(*byte as char);
    }
    out
}

fn format_absolute(value: f64, min_fraction: u32, max_fraction: u32, spec: &LocaleSpec) -> String {
    let rounded = round_half_away_from_zero(value, max_fraction);
    let integer_part = rounded.trunc() as u128;
    let integer_str = insert_thousands(integer_part, spec.thousands);

    if max_fraction == 0 {
        return integer_str;
    }

    let remainder = (rounded - rounded.trunc()).abs();
    // `format!("{:.N}", x)` already rounds to N digits; because `rounded` was
    // snapped to max_fraction digits above, this formatting simply extracts
    // those digits without re-rounding in a lossy direction.
    let formatted_remainder = format!("{:.*}", max_fraction as usize, remainder);
    let fraction_digits = formatted_remainder
        .strip_prefix("0.")
        .unwrap_or(&formatted_remainder);

    let trimmed = fraction_digits.trim_end_matches('0');
    let min = min_fraction as usize;
    let final_fraction: String = if trimmed.len() < min {
        fraction_digits.chars().take(min).collect()
    } else {
        trimmed.to_string()
    };

    if final_fraction.is_empty() {
        integer_str
    } else {
        format!("{}{}{}", integer_str, spec.decimal, final_fraction)
    }
}

/// Formats `value` according to `options`, mirroring `Intl.NumberFormat`.
///
/// Returns the literal string `"NaN"` for NaN input, matching the TypeScript
/// reference and the Python port.
///
/// # Arguments
///
/// * `value` - The number to format.
/// * `options` - Locale, style, currency, and fraction-digit configuration.
///
/// # Returns
///
/// The formatted number as a `String`.
///
/// # Examples
///
/// ```
/// use umt_rust::number::{umt_format_number, FormatNumberOptions, FormatStyle};
///
/// assert_eq!(
///     umt_format_number(1234567.89, &FormatNumberOptions::default()),
///     "1,234,567.89",
/// );
///
/// let de = FormatNumberOptions {
///     locale: Some("de-DE"),
///     ..FormatNumberOptions::default()
/// };
/// assert_eq!(umt_format_number(1234567.89, &de), "1.234.567,89");
///
/// let percent = FormatNumberOptions {
///     style: Some(FormatStyle::Percent),
///     ..FormatNumberOptions::default()
/// };
/// assert_eq!(umt_format_number(0.75, &percent), "75%");
///
/// let usd = FormatNumberOptions {
///     style: Some(FormatStyle::Currency),
///     currency: Some("USD"),
///     locale: Some("en-US"),
///     ..FormatNumberOptions::default()
/// };
/// assert_eq!(umt_format_number(1234.5, &usd), "$1,234.50");
/// ```
pub fn umt_format_number(value: f64, options: &FormatNumberOptions) -> String {
    if value.is_nan() {
        return "NaN".to_string();
    }

    let style = options.style.unwrap_or_default();
    let (default_min, default_max) = default_fraction_digits(style, options.currency);
    let min_fraction = options.minimum_fraction_digits.unwrap_or(default_min);
    // Intl.NumberFormat forgives max < min by lifting max up to min.
    let max_fraction = options
        .maximum_fraction_digits
        .unwrap_or(default_max)
        .max(min_fraction);

    let spec = resolve_locale_spec(options.locale);

    let scaled_value = if style == FormatStyle::Percent {
        value * 100.0
    } else {
        value
    };

    let sign = if scaled_value.is_sign_negative() && scaled_value != 0.0 {
        "-"
    } else {
        ""
    };
    let magnitude = format_absolute(scaled_value.abs(), min_fraction, max_fraction, &spec);

    match style {
        FormatStyle::Percent => format!("{}{}%", sign, magnitude),
        FormatStyle::Currency => {
            let symbol = match options.currency {
                Some(code) => match currency_symbol(code) {
                    Some(sym) => sym.to_string(),
                    // ISO fallback uses a non-breaking space, as Intl does.
                    None => format!("{}\u{00a0}", code),
                },
                None => String::new(),
            };
            format!("{}{}{}", sign, symbol, magnitude)
        }
        FormatStyle::Decimal => format!("{}{}", sign, magnitude),
    }
}

/// Convenience wrapper that formats with all defaults (decimal, en-US layout).
///
/// # Arguments
///
/// * `value` - The number to format.
///
/// # Returns
///
/// The formatted number as a `String`.
pub fn umt_format_number_default(value: f64) -> String {
    umt_format_number(value, &FormatNumberOptions::default())
}
